rfcs/20220404-troubleshooting-helpers/README.md

```diff
@@ -0,0 +1,304 @@
+
+## A common interface to log messages
+
+This could be achieved by providing a set of macros that can be used
+within implementations. We chose macros as it allows to avoid
+boilerplate code required to pass common information explicitly as
+parameters (e.g. filename, line number, `pd()->info()` for dispatch
+info, ...). Inside the macro call, we can still rely on functions to
+avoid the library size from blowing up.
+
+Those macros should not take strings as input except for
+implementation specific information for developers. For common
+messages, we should use enum values as it has 2 main benefits:
+- makes common messages map to the same string, which simplifies
+  parsing the verbose output to find an information, as there is an
+  exact match between printed string and the kind of message it
+  carries.
+- simplifies potential future localization efforts.
+
+
+Here is the list of macros that were identified as potential
+candidates.  First check functions targeted at displaying
+errors. Those should return a value when some condition is false and
+they should typically be used when we return from an user visible
+function.
+- `CHECK(cat, cond, ret_val, msg_enum)`, will check the condition
+  `cond`, emit the message according to the category of messages
+  `cat`, and return `ret_val`. This macro would replace the existing
+  `CHECK` macro.
+- `<THIRDPARTY>_CHECK(msg_cat, cond, ret_val, msg_enum)` would allow to report
+  failed checks from calls to third party dependencies
+  (e.g. `SYCL_CHECK`, `OCL_CHECK`, `ACL_CHECK`, ...)
+
+Another set of macro to display other kind of user oriented information:
+- `PROFILE(msg_cat, )` These are mainly for printing creation and
+  execution time.
+- `DISPATCH_CHECK(cond, msg_enum)` would print msg in the log, with
+  proper formatting. This macro is to be used in init function of each
+  primitive implementation. The granularity is implementation defined,
+  but general messaging should be enough (e.g. if only trivial strides
+  / dense tensors are supported for an implementation, no need to
+  specify which input does not satisfy the condition).  If the
+  condition is false, this macro returns `unimplemented`.  This would
+  be a syntactic sugar macro for `CHECK(dispatch, cond,
+  status::unimplemented, msg_enum)`
+- `ADVICE(cond, msg_enum)` this is for cases where we do dispatch our
+  best performing implementation, but we can provide advice to user
+  that they can change something to make it faster. A typical example
+  is to let user know that they use memory with sub-optimal alignment,
+  or in the case of Matmul, they could use different leading
+  dimensions to improve performance.
+
+A set of macros that are prefixed with `D`, which would be enabled
+only when the library is built with `ONEDNN_DEVEL_MODE=ON`. Those
+should be compiled away when the library is built with
+`ONEDNN_DEVEL_MODE=OFF` (default).
+- `DINFO(level, get_string)` to print developer info. Here
```

Comment:
As an option we could rely on a stream-like interface (to avoid an extra step with lambda when a message can be put in-place). For example:

```c++
DINFO(level) << "Parameter X = " << X << std::endl;
```

https://github.com/oneapi-src/oneDNN/blob/11fa74eaf03af9848c1bb5fffb4cbb2866aadf42/src/gpu/jit/conv/kernel_builder.cpp#L6596-L6597